geo/src/algorithm/geodesic_area.rs

```diff
@@ -0,0 +1,504 @@
+    /// 
+    /// # Area Assumptions
+    ///  - Polygons are assumed to be wound in a counter-clockwise direction 
+    ///    for the exterior ring and a clockwise direction for interior rings. 
+    ///    This is the standard winding for Geometries that follow the Simple Features standard. 
+    ///    Using alternative windings will result in incorrect results.
+    /// 
+    /// # Perimeter
+    /// For a polygon this returns the perimeter of the exterior ring and interior rings.
+    /// To get the perimeter of just the exterior ring of a polygon, do `polygon.exterior().geodesic_length()`.
+    ///
+    /// # Units
+    ///
+    /// - return value: (meter, meter²)
+    /// 
+    /// [Karney (2013)]:  https://arxiv.org/pdf/1109.4448.pdf
+    fn geodesic_perimeter_area_unsigned(&self) -> (T, T);
+}
+
+impl GeodesicArea<f64> for Polygon {
+    fn geodesic_perimeter(&self) -> f64 {
+        let (perimeter, _area) = geodesic_area(self, true, false, false);
+        perimeter
+    }
+
+    fn geodesic_area_signed(&self) -> f64 {
+        let (_perimeter, area) = geodesic_area(self, true, false, false);
+        area
+    }
+
+    fn geodesic_area_unsigned(&self) -> f64 {
+        let (_perimeter, area) = geodesic_area(self, false, false, false);
+        area
+    }
+
+    fn geodesic_perimeter_area_signed(&self) -> (f64, f64) {
+        geodesic_area(self, true, false, false)
+    }
+
+    fn geodesic_perimeter_area_unsigned(&self) -> (f64, f64) {
+        geodesic_area(self, false, false, false)
+    }
+}
+
+fn geodesic_area(poly: &Polygon, sign: bool, reverse: bool, exterior_only: bool) -> (f64, f64) {
+    let g = Geodesic::wgs84();
+
+    let (exterior_winding, interior_winding) = if reverse {
+        (Winding::Clockwise, Winding::CounterClockwise)
+    } else {
+        (Winding::CounterClockwise, Winding::Clockwise)
+    };
+
+    // Add the exterior ring
+    let (outer_perimeter, outer_area) = {
+        let mut pa = PolygonArea::new(&g, exterior_winding);
+        poly.exterior().points().for_each(|p| {
+            pa.add_point(p.y(), p.x());
+        });
+        let (perimeter, area, _) = pa.compute(sign);
+        (perimeter, area)
+    };
+
+    // Add the interior rings
+    let (interior_perimeter, mut inner_area) = if exterior_only {
+        (0.0, 0.0)
+    } else {
+        let mut inner_area = 0.;
+        let mut inner_perimeter = 0.;
+        poly.interiors().iter().for_each(|ring| {
+            let mut pa = PolygonArea::new(&g, interior_winding);
+            ring.points().for_each(|p| {
+                pa.add_point(p.y(), p.x());
+            });
+            let (perimeter, area, _) = pa.compute(sign);
+            inner_area += area.abs();
+            inner_perimeter += perimeter;
+        });
+        (inner_perimeter, inner_area)
+    };
+
+    if outer_area < 0.0 && inner_area > 0.0 {
```

Comment:
Why this check? It seems like you might be trying to "fix" someone's winding for them?

Reply:
This is a best-effort fix for badly wound polygons.  I found this looking at the test suite for `ChamberlainDuquetteArea`. The test-suite contains a badly-wound polygon here (https://github.com/georust/geo/blob/main/geo/src/algorithm/chamberlain_duquette_area.rs#L228).  Both the exerior-ring and interior-rings are wound counter-clockwise, which is incorrect as per Simple Features spec.   However, we can easily detect this and correct for it, which is what you see here.

Reply:
It might make sense to add a bunch more test to ensure that this logic isn't having any unintended consequences.  I'd be happy to do so if you think it useful.